flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/window/utils/WindowJoinHelper.java

```diff
@@ -0,0 +1,403 @@
+
+    private final WindowTimerService<Long> windowTimerService;
+
+    protected final RowDataSerializer leftSerializer;
+
+    protected final RowDataSerializer rightSerializer;
+
+    protected final JoinConditionWithNullFilters joinCondition;
+
+    private final int leftWindowEndIndex;
+    private final int rightWindowEndIndex;
+
+    /** This is used for emitting elements with a given timestamp. */
+    protected final TimestampedCollector<RowData> collector;
+
+    private final WindowJoinProcessor windowJoinProcessor;
+
+    // ------------------------------------------------------------------------
+    // Metrics
+    // ------------------------------------------------------------------------
+
+    private Meter leftLateRecordsDroppedRate;
+    private Meter rightLateRecordsDroppedRate;
+
+    public WindowJoinHelper(
+            RowDataSerializer leftSerializer,
+            RowDataSerializer rightSerializer,
+            ZoneId shiftTimeZone,
+            WindowTimerService<Long> windowTimerService,
+            JoinConditionWithNullFilters joinCondition,
+            int leftWindowEndIndex,
+            int rightWindowEndIndex,
+            TimestampedCollector<RowData> collector,
+            FlinkJoinType joinType) {
+        this.leftSerializer = leftSerializer;
+        this.rightSerializer = rightSerializer;
+        this.shiftTimeZone = shiftTimeZone;
+        this.windowTimerService = windowTimerService;
+        this.joinCondition = joinCondition;
+        this.leftWindowEndIndex = leftWindowEndIndex;
+        this.rightWindowEndIndex = rightWindowEndIndex;
+        this.collector = collector;
+
+        this.windowJoinProcessor = getWindowJoinProcessor(joinType);
+    }
+
+    public void registerMetric(OperatorMetricGroup metrics) {
+        // register metrics
+        Counter leftNumLateRecordsDropped = metrics.counter(LEFT_LATE_ELEMENTS_DROPPED_METRIC_NAME);
+        this.leftLateRecordsDroppedRate =
+                metrics.meter(
+                        LEFT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME,
+                        new MeterView(leftNumLateRecordsDropped));
+        Counter rightNumLateRecordsDropped =
+                metrics.counter(RIGHT_LATE_ELEMENTS_DROPPED_METRIC_NAME);
+        this.rightLateRecordsDroppedRate =
+                metrics.meter(
+                        RIGHT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME,
+                        new MeterView(rightNumLateRecordsDropped));
+        metrics.gauge(
+                WATERMARK_LATENCY_METRIC_NAME,
+                () -> {
+                    long watermark = windowTimerService.currentWatermark();
+                    if (watermark < 0) {
+                        return 0L;
+                    } else {
+                        return windowTimerService.currentProcessingTime() - watermark;
+                    }
+                });
+    }
+
+    public void processElement(RowData inputRow, boolean isLeft) throws Exception {
```

Comment:
Compare with the original one in `WindowJoinOperator`, add a new boolean flag isLeft seems to make the code a bit more complicated, WDYT?

Reply:
Done for it.